ethportal-api/src/types/execution/transaction.rs

```diff
@@ -71,103 +71,59 @@ impl Transaction {
                 tx.encode(out);
             }
             Self::Blob(tx) => {
-                let payload_length = tx.fields_len();
-                if with_header {
-                    RlpHeader {
-                        list: false,
-                        payload_length: 1 + length_of_length(payload_length) + payload_length,
-                    }
-                    .encode(out);
-                }
                 out.put_u8(TransactionId::Blob as u8);
                 tx.encode(out);
             }
         }
     }
 
-    pub fn decode_enveloped_transactions(buf: &mut &[u8]) -> alloy::rlp::Result<Self> {
-        // at least one byte needs to be present
-        if buf.is_empty() {
-            return Err(RlpError::InputTooShort);
-        }
-        let original_encoding = *buf;
-        let header = RlpHeader::decode(buf)?;
-        let value = &mut &buf[..header.payload_length];
-        buf.advance(header.payload_length);
-        if !header.list {
-            let id = TransactionId::try_from(value[0])
-                .map_err(|_| RlpError::Custom("Unknown transaction id"))?;
-            value.advance(1);
-            match id {
-                TransactionId::EIP1559 => Ok(Self::EIP1559(EIP1559Transaction::decode(value)?)),
-                TransactionId::AccessList => {
-                    Ok(Self::AccessList(AccessListTransaction::decode(value)?))
-                }
-                TransactionId::Legacy => {
-                    unreachable!("Legacy transactions should be wrapped in a list")
-                }
-                TransactionId::Blob => Ok(Self::Blob(BlobTransaction::decode(value)?)),
-            }
-        } else {
-            Ok(Self::Legacy(LegacyTransaction::decode(
-                &mut &original_encoding[..(header.payload_length + header.length())],
-            )?))
+    fn length(&self) -> usize {
+        match self {
+            Self::Legacy(tx) => tx.length(),
+            Self::AccessList(tx) => 1 + tx.length(),
+            Self::EIP1559(tx) => 1 + tx.length(),
+            Self::Blob(tx) => 1 + tx.length(),
         }
     }
 }
 
-impl Encodable for Transaction {
-    fn encode(&self, out: &mut dyn bytes::BufMut) {
-        self.encode_with_envelope(out, false)
-    }
-}
-
 impl Decodable for Transaction {
-    fn decode(buf: &mut &[u8]) -> alloy::rlp::Result<Self> {
+    fn decode(buf: &mut &[u8]) -> rlp::Result<Self> {
         // at least one byte needs to be present
         if buf.is_empty() {
-            return Err(RlpError::InputTooShort);
+            return Err(rlp::Error::InputTooShort);
         }
         let id = TransactionId::try_from(buf[0])
-            .map_err(|_| RlpError::Custom("Unknown transaction id"))?;
+            .map_err(|_| rlp::Error::Custom("Unknown transaction id"))?;
         match id {
+            TransactionId::Blob => {
+                buf.advance(1);
+                BlobTransaction::decode(buf).map(Self::Blob)
```

Comment:
Is there any reason besides stylistic preference to use `buf.advance(1)` vs passing in `&buf[1..]`?

I'm pretty neutral on the choice, just curious if I missed anything.

Reply:
Actually, yes there is. Original implementation actually doesn't work correctly according to spec.

The documentation of the [Decodable::decode](https://docs.rs/alloy-rlp/0.3.8/alloy_rlp/trait.Decodable.html#tymethod.decode) says: _`buf` must be advanced past the decoded object._

Passing `&mut &buf[1..]` would not modify `buf` at all. That's why we have to:

1. advance `buf` by 1 to skip transaction type
2. pass `buf` into `*Transaction::decode` which will update `buf` as it decodes the transaction

---

The "legacy transaction" is different in this regard, as it doesn't use 1 byte for transaction type (legacy reasons), but it's first byte is clearly different than any of the non-legacy types (this is handled by code in `TransactionId::try_from`).
So decoding logic is to peek at the first byte, figure out if it is legacy or not, and proceed accordingly.